sel/src/Sel/Hashing/Generic.hs

```diff
@@ -0,0 +1,171 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+
+-- |
+--
+-- Module: Sel.Hashing.Generic
+-- Description: Fingerprint hashing with the BLAKE2b algorithm
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Portability: GHC only
+module Sel.Hashing.Generic
+  ( -- * Introduction
+    -- $introduction
+
+    -- * Operations
+    HashKey
+  , newHashKey
+  , Hash
+  , hashByteString
+  , hashToText
+  , hashToByteString
+  , hashToBinary
+  )
+where
+
+import Control.Monad (void)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Base16 as Base16
+import Data.ByteString.Unsafe (unsafePackMallocCStringLen, unsafeUseAsCStringLen)
+import Data.Text (Text)
+import Data.Text.Display
+import qualified Data.Text.Lazy.Builder as Builder
+import Foreign (ForeignPtr, Ptr)
+import qualified Foreign
+import Foreign.C (CChar, CSize, CUChar)
+import Foreign.Storable
+import GHC.IO.Handle.Text (memcpy)
+import LibSodium.Bindings.GenericHashing (cryptoGenericHash, cryptoGenericHashBytes, cryptoGenericHashKeyBytes, cryptoGenericHashKeyGen)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+-- $introduction
+--
+-- This API computes a fixed-length fingerprint for an arbitrarily long message.
+-- It is backed by the BLAKE2b algorithm.
+--
+-- Sample use cases:
+--
+--   * File integrity checking
+--   * Creating unique identifiers to index arbitrarily long data
+--
+-- __⚠️ Do not use this module to hash passwords! ⚠️__
+--
+-- If you need to deviate from the defaults enforced by this module,
+-- please use the underlying bindings at "LibSodium.Bindings.GenericHashing".
+
+-- |
+--
+-- @since 0.0.1.0
+newtype HashKey = HashKey (ForeignPtr CUChar)
+  deriving newtype
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    )
+
+-- | Create a new 'HashKey' of size 'cryptoGenericHashKeyBytes'.
+--
+-- @since 0.0.1.0
+newHashKey :: IO HashKey
+newHashKey = do
+  fPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoGenericHashKeyBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoGenericHashKeyGen ptr
+  pure $ HashKey fPtr
+
+-- | The fingerprint computed by `hashByteString`.
+-- It is produced by the BLAKE2b algorithm, and is
+-- of size 'cryptoGenericHashBytes', as recommended.
+--
+-- You can produce a human-readable string representation
+-- of a 'Hash' by using the `display` function.
+--
+-- @since 0.0.1.0
+newtype Hash = Hash (ForeignPtr CUChar)
+  deriving newtype
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    )
+
+instance Storable Hash where
+  sizeOf :: Hash -> Int
+  sizeOf _ = sizeOf (undefined :: CUChar)
```

Comment:
Are hashes one byte wide?

Reply:
They can be at most `cryptoGenericHashBytesMax` (64), so I guess I could use CSize? But I don't see a reason why I shouldn't and this bugs me off.

Reply:
So is `Hash` variable-width then?